storage/index.md

```diff
@@ -70,38 +70,117 @@ PostgreSQL have some multi-modality capability.]
+a query that's slow, you can try to write out or re-materialize the
+out with a different field as the ordering constraint.
+
+In all databases, query optimization is often about finding ways to
+limit the amount of data to consider. In record based systems,
+normalization (splitting logical records across tables,) and indexes
+(materializing compact views) reduces the amount of data to
+consider. For column based systems, saving additional copies of data
+in different forms with different orderings, or that contain a subset
+of fields, can have a huge impact on query performance.
+
+<!-- TODO: connect to "use COPY TO to materialize different views of your data" -->
 
 ### Storage Engines and Encoding
 
-- discuss MVCC and implications for GlareDB, 
-- mention vacuuming requirements
-- discuss encoding efficencies (protobufs and bson)
+GlareDB's data sources can be classified in a few groups:
+
+- structured and semi-structured file formats. There are many
+  different interchange formats, record encoding, and archival formats
+  that allow us to write data to files, and share them between
+  programs. JSON, Parquet, CSV, BSON, and Excel are just encoded raw
+  data formats that GlareDB can interpret and present as a table. Each
+  format has different properties and can provide different benefits
+  to users. Some are easier to write and more widely supported as
+  output formats, some are easier or more efficient to read
+  programtiacally, and some are more transparent to read for humans.
+
+- external database engines and services: GlareDB provides access to
+  other database systems--like PostgreSQL, MySQL, MongoDB, SQLite,
+  Snowflake and BigQuery, to enable joining data across data sources,
+  or migrating data between formats and systems. Most of the time
+  GlareDB is able to push operations _into_ the underlying database
+  engine, which means these queries can take advantage of indexes in
+  the underlying data and reduce the amount of data it needs to
+  consider.
+
+- storage engines, like Lance, DeltaLake and Iceberg, provide
+  transactional and consistently properties that file formats cannot
+  without the overhead of a process. At the same time, unlike extenral
+  services and engines, these formats don't require a controling
+  process and can work with all of the data stored to an object store
+  like S3 or GCS. They do this by defining a protocol for writing and
+  reading files that ensure that data is consistent and that writes
+  are safe.
+
+There are exceptions and edge cases: SQLite is somewhere between a
+database and a file format, and the Parquet format has metadata that
+makes it possible to push down some operations more like an external
+database, while the implementations of Delta Lake are built upon
+Parquet and JSON. However, this classification system may be useful
+in understanding the limitations and features of a specific data
+system.
 
 ### Materialization: Roll Ups and Duplication
```

Comment:
I'm unsure of the capitalization, but pretty sure this should be hyphenated as a noun.
```suggestion
### Materialization: Roll-ups and Duplication
```